Migrating a subset of Xtend code from org.eclipse.xtext.xbase.ui project to Java

nbhusare opened the pull request:
- [x] Migrated subset of Xtend code to Java
- [x] Ran the Maven/Gradle build locally

Signed-off-by: nbhusare <neerajbhusare@gmail.com>

nbhusare commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java:
```diff
@@ -0,0 +1,162 @@
+								IPath relative = resource.getFullPath().removeFirstSegments(sourceFolder.getOutputLocation().segmentCount());
+								Optional<IPackageFragmentRoot> source = FluentIterable.from(project.findPackageFragmentRoots(sourceFolder))
+										.first();
+								if (source.isPresent()) {
```
Added a safety-check which was not present in the original code. The code `project.findPackageFragmentRoots(sourceFolder)` may return an empty array.

tivervac replied:
Seems fine to me especially with the extra log

nbhusare commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java:
```diff
@@ -0,0 +1,162 @@
+				}
+			}
+		} catch (JavaModelException e) {
+			LOG.error("Error finding the original source for the editor input", e);
```
Logging an error message. The chose to do it after looking at other places in the source code where `JavaModelException` is caught. Please suggest if you would like me to replace with a `RuntimeException`.

tivervac replied:
Going to leave this one for @szarnekow

szarnekow replied:
Please only log on error level if the exception is not `e.isDoesNotExist()`

nbhusare replied:
@szarnekow What should I do if the condition `e.isDoesNotExist()` is not met?

nbhusare commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java:
```diff
@@ -0,0 +1,162 @@
+				return input;
+			}
+
+			Optional<? extends ILocationInEclipseResource> sourceInformation = Iterables.tryFind(trace.getAllAssociatedLocations(),
```
Simplified the code. The original code made use of a `while-loop` to find the first element for which the “languageInfo.equals(candidate.getLanguage())” is true. I do the same using the new code

nbhusare commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java:
```diff
@@ -0,0 +1,162 @@
+	 *             if the provided {@code element} is null
+	 */
+	public IPackageFragmentRoot getPackageFragmentRoot(IJavaElement element) {
+		Preconditions.checkNotNull(element, "Couldn't find source package fragment root.");
```
Added a safety-check here.

tivervac replied:
LGTM

nbhusare commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/copyqualifiedname/XbaseCopyQualifiedNameService.java:
```diff
@@ -0,0 +1,197 @@
+ * @author Anton Kosyakov - Initial contribution and API
+ * @since 2.4
+ */
+public class XbaseCopyQualifiedNameService extends DefaultCopyQualifiedNameService {
```
`org.eclipse.xtext.ui.editor.copyqualifiedname.DefaultCopyQualifiedNameService` is deprecated. We'll have to revisit this code and replace the deprecated API with the correct ones. Corresponding change needs to be made in the `XbaseUiModule#bindCopyQualifiedNameService()` method.

szarnekow replied:
Please take a look at the conversion that was applied to the org.eclipse.xtext.naming.DefaultCopyQualifiedNameService - there we use a PolymorphicDispatcher. 
I'm afraid we have to do the same here in order to remain backwards compatible.

nbhusare replied:
I'll file a separate jira for this.

szarnekow replied:
https://github.com/eclipse/xtext-extras/issues/581

nbhusare commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/copyqualifiedname/XbaseCopyQualifiedNameService.java:
```diff
@@ -0,0 +1,197 @@
+			declaredConstructors = resolvedFeatures.getDeclaredConstructors();
+		}
+
+		Optional<IResolvedConstructor> resolvedConstructor = FluentIterable.from(declaredConstructors)
```
`FluentIterable` - A bit of inspiration from @szarnekow code.

tivervac replied:
Does this work if declaredConstructors is null?

nbhusare replied:
`declaredConstructors` can be empty, but not null. PS - https://bit.ly/3dIG53z

Generally speaking, `FluentIterable.from` will throw a NPE if the passed param. is null.

szarnekow replied:
declaredConstructors will still be empty if resolvedFeatures is null.

These null checks look a little too xtendy to me. Many of the intermediate local vars will never be null if the actual receiver was present.
Please reverse the logic and exit early but otherwise keep a normal flow.

szarnekow replied:
Also my preference would always be something like
```
for(IResolvedConstructor candidate: declaredConstructors) {
  if (constructor.equals(candidate.getDeclaration()) {
    return candidate;
 }
}
return null

```

tivervac replied:
It is going to be null if `resolvedFeatures` is null

nbhusare replied:
You are right, I missed that.

nbhusare commented:
@cdietrich @tivervac @szarnekow Kindly review.

tivervac commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/contentassist/XbaseContentProposalPriorities.java:
```diff
@@ -0,0 +1,52 @@
+			Object desc = configurableProposal.getAdditionalData(XbaseProposalProvider.DESCRIPTION_KEY);
+			if (desc instanceof SimpleIdentifiableElementDescription) {
+				if (!("this".equals(configurableProposal.getReplacementString())
+						&& "super".equals(configurableProposal.getReplacementString()))) {
```
Either add the `!` before both equals, or do `&&` -> `||`

tivervac commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java:
```diff
@@ -0,0 +1,162 @@
+
+	public IEditorInput findOriginalSource(IEditorInput input) {
+		IFile resource = ResourceUtil.getFile(input);
+		if (resource != null) {
```
Might be better to just write `if (resource == null) return input;`

tivervac commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/copyqualifiedname/XbaseCopyQualifiedNameService.java:
```diff
@@ -0,0 +1,197 @@
+		} else if (constructor == null && constructorCall == null) {
+			return _getQualifiedName((Void) null, (Void) null);
+		} else {
+			throw new IllegalArgumentException("Unhandled parameter types: " + Arrays.asList(constructor, constructorCall).toString());
```
You can remove this, it's impossible to end up here, I don't if you leave it though

tivervac commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/copyqualifiedname/XbaseCopyQualifiedNameService.java:
```diff
@@ -0,0 +1,197 @@
+		}
+
+		Optional<IResolvedConstructor> resolvedConstructor = FluentIterable.from(declaredConstructors)
+				.filter(constr -> constr.getDeclaration().equals(consturctor)) //
```
constructor*

nbhusare replied:
Let me change it to `declaredConstructor` to be consistent with the list `declaredConstructors`.

szarnekow commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/contentassist/XbaseContentProposalPriorities.java:
```diff
@@ -0,0 +1,52 @@
+					return;
+				}
+			} else if (desc instanceof StaticFeatureDescriptionWithTypeLiteralReceiver) {
+				adjustPriority(proposal, prefix, 560);
```
This will fall through to the super.adjust.. call. Not sure if that was the intention in the xtend code.

nbhusare replied:
PS - https://bit.ly/3bJ43Kg

szarnekow replied:
I think this was a bug in the old impl.

szarnekow commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java:
```diff
@@ -0,0 +1,165 @@
+	 */
+	public IPackageFragmentRoot getPackageFragmentRoot(IJavaElement element) {
+		Preconditions.checkNotNull(element, "Couldn't find source package fragment root.");
+		return element instanceof IPackageFragmentRoot ? (IPackageFragmentRoot) element : getPackageFragmentRoot(element.getParent());
```
We probably want to use `org.eclipse.jdt.core.IJavaElement.getAncestor(int)` here

nbhusare replied:
Very good suggestion. Tx.

szarnekow commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java:
```diff
@@ -0,0 +1,165 @@
+						for (IClasspathEntry sourceFolder : classpathEntries) {
+							if (sourceFolder.getOutputLocation() != null && sourceFolder.getOutputLocation().isPrefixOf(resource.getFullPath())) {
+								IPath relative = resource.getFullPath().removeFirstSegments(sourceFolder.getOutputLocation().segmentCount());
+								Optional<IPackageFragmentRoot> source = FluentIterable.from(project.findPackageFragmentRoots(sourceFolder))
```
Iff we change this code, we could loop over all matching packageFragmentRoots until you find one that fulfills `newFile.exists()` below.

szarnekow commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java:
```diff
@@ -0,0 +1,165 @@
+						}
+
+						// Check if it's sitting in one of the output folders set on the source folders
+						List<IClasspathEntry> classpathEntries = new ArrayList<>();
```
Populating a list to loop over the list on step later appears to be unnecessary. Maybe extract a few smaller methods to improve the readability and avoid the intermediate list here?

szarnekow commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java:
```diff
@@ -0,0 +1,165 @@
+						if (project.getOutputLocation().isPrefixOf(resource.getFullPath())) {
+							IPath relative = resource.getFullPath().removeFirstSegments(project.getOutputLocation().segmentCount());
+
+							List<IPackageFragmentRoot> packageFragmentRoots = new ArrayList<>();
```
Same comment here as a few lines below: We create unnecessary lists and end up with a very long method. Maybe it's better to refactor this by extracting a few smaller methods.

szarnekow commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java:
```diff
@@ -0,0 +1,165 @@
+			return input;
+		}
+
+		Optional<? extends ILocationInEclipseResource> sourceInformation = Iterables.tryFind(trace.getAllAssociatedLocations(),
```
I'd prefer a more straight forward 
```
for(PlatformResource candidate: trace.getAllAssociatedLocations()) {
  if (languageInfo.equals(candidate.getLanguage()) {
    IStorage storage = candidate.getPlatformResource();
    if (storage != null) {
      return ..
    }
  }
}
return input;
```

nbhusare replied:
The above code goes through all-the-candidates until it finds one for which `storage != null`. If you look at the original code (https://bit.ly/2wXAeXN), it gets the **first** candidate that matches the condition `languageInfo.equals(candidate.getLanguage())`, and uses it for further processing. 

@szarnekow Is it possible for more than one `candidate` to match the condition `languageInfo.equals(candidate.getLanguage())` ?

szarnekow commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/copyqualifiedname/XbaseCopyQualifiedNameService.java:
```diff
@@ -0,0 +1,195 @@
+		Optional<IResolvedConstructor> resolvedConstrutor = FluentIterable.from(declaredConstructors)
+				.filter(constr -> constr.getDeclaration().equals(constructor)) //
+				.first();
+		return resolvedConstrutor.isPresent() ? resolvedConstrutor.get() : null;
```
Typo in variable name

szarnekow commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/copyqualifiedname/XbaseCopyQualifiedNameService.java:
```diff
@@ -0,0 +1,195 @@
+		}
+	}
+
+	protected IResolvedExecutable _resolveExecutable(JvmConstructor consturctor, XAbstractFeatureCall featureCall,
```
Typo in parameter name

szarnekow commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/copyqualifiedname/XbaseCopyQualifiedNameService.java:
```diff
@@ -0,0 +1,195 @@
+			allOperations = resolvedFeatures.getAllOperations();
+		}
+
+		Optional<IResolvedOperation> resolvedOperation = FluentIterable.from(allOperations)
```
Not sure how often we do similar things in the code base, but I can imagine this logic to be moved to `ResolvedFeatures`, e.g. `findResolvedOperation(JvmOperation declaration)`

nbhusare replied:
The methods `ResolvedFeatures#getAllOperations()` and `ResolvedFeatures#getDeclaredConstructors()` are used in very few places, not for the same purpose as we have here. Nevertheless, I like you idea of adding methods `#findResolvedOperation(JvmOperation declaration)` (and  `#findDeclaredConstructor(JvmConstructor constructor)`) to the `ResolvedFeatures` class.

szarnekow commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/util/InsertionOffsets.java:
```diff
@@ -0,0 +1,41 @@
+
+	public int inEmpty(EObject element) {
+		ICompositeNode compositeNode = NodeModelUtils.findActualNodeFor(element);
+		Optional<ILeafNode> openingBraceNode = Iterables.tryFind(compositeNode.getLeafNodes(), (node) -> "{".equals(node.getText()));
```
Not a big fan of these usages of optionals. Checking for `isPresent` is usually not good style. You probably want to do something like `Iterables.tryFind(..).transform(node->getOffset() + 1).or(compositeNode::getEndOffset)`

szarnekow replied:
Or preferably a plain for loop.

nbhusare commented:
@szarnekow @tivervac Tx for the review. I have addressed all your review comments. Kindly have a look, and resolve the conversation if you are ok with the changes.

szarnekow commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java:
```diff
@@ -0,0 +1,150 @@
+				}
+			}
+		} catch (JavaModelException e) {
+			if (e.isDoesNotExist()) {
```
It's the opposite of what I tried to suggest: Only log, ig it is not `e.isDoesNotExist()`

szarnekow commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java:
```diff
@@ -0,0 +1,150 @@
+	 */
+	public IPackageFragmentRoot getPackageFragmentRoot(IJavaElement element) {
+		Preconditions.checkNotNull(element, "Couldn't find source package fragment root.");
+		return element instanceof IPackageFragmentRoot //
```
The semantics of getAncestor do already imply the check of the element itself. Can be implemented as `return (IPackageFragmentRoot) element.getAncestor()`

szarnekow commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/copyqualifiedname/XbaseCopyQualifiedNameService.java:
```diff
@@ -0,0 +1,172 @@
+		XExpression actualReceiver = featureCall.getActualReceiver();
+		if (actualReceiver != null) {
+			LightweightTypeReference actualType = resolvedTypes.getActualType(actualReceiver);
+			ResolvedFeatures resolvedFeatures = overrideHelper.getResolvedFeatures(actualType);
```
Code is duplicated in the method below.

tivervac commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java:
```diff
@@ -0,0 +1,150 @@
+	public IEditorInput findOriginalSourceForOuputFolderCopy(IEditorInput input) {
+		try {
+			IFile resource = ResourceUtil.getFile(input);
+			if (resource != null) {
```
I would early return here as well, also this doesn't need to be in try catch, try to keep it as small as possible

tivervac commented on org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java:
```diff
@@ -0,0 +1,150 @@
+		try {
+			IFile resource = ResourceUtil.getFile(input);
+			if (resource != null) {
+				// if the given resource is already pointing to a language file
```
s/if/If/

nbhusare commented:
@tivervac @szarnekow Addressed review comments. Please check.

cdietrich commented:
many thanks @nbhusare
